src/wp-admin/js/editor.js

```diff
@@ -116,210 +202,327 @@
+			 * followed by the captured block level element.
+			 */
 			text = text.replace( new RegExp( '(<(?:' + blocklist + ')(?: [^>]*)?>)', 'gi' ), '\n$1' );
+			/*
+			 * Replaces a block level element closing tag with the captured
+			 * block level element followed by 2 newlines.
+			 */
 			text = text.replace( new RegExp( '(</(?:' + blocklist + ')>)', 'gi' ), '$1\n\n' );
-			text = text.replace( /<hr( [^>]*)?>/gi, '<hr$1>\n\n' ); // hr is self closing block element
-			text = text.replace( /\s*<option/gi, '<option' ); // No <p> or <br> around <option>
+			// Adds 2 newlines to a HR tag.
+			text = text.replace( /<hr( [^>]*)?>/gi, '<hr$1>\n\n' );
+			/*
+			 * Replaces an option open tag preceded by spaces
+			 * with an option tag without spaces.
+			 */
+			text = text.replace( /\s*<option/gi, '<option' );
+			/*
+			 * Replaces an option end tag followed by spaces
+			 * with an option tag without spaces.
+			 */
 			text = text.replace( /<\/option>\s*/gi, '</option>' );
+			/*
+			 * Replaces a newline followed by one or multiple spaces
+			 * and another newline with 2 newlines.
+			 */
 			text = text.replace( /\n\s*\n+/g, '\n\n' );
+			// Replaces 2 newlines with a paragraph and a single newline.
 			text = text.replace( /([\s\S]+?)\n\n/g, '<p>$1</p>\n' );
+			// Replaces p tags containing only white spaces with nothing.
 			text = text.replace( /<p>\s*?<\/p>/gi, '');
+			// Removes spaces and p tags around block level elements.
 			text = text.replace( new RegExp( '<p>\\s*(</?(?:' + blocklist + ')(?: [^>]*)?>)\\s*</p>', 'gi' ), '$1' );
+			// Removes p tags around li elements.
 			text = text.replace( /<p>(<li.+?)<\/p>/gi, '$1');
+			// Removes spaces and p tags from blockquotes.
 			text = text.replace( /<p>\s*<blockquote([^>]*)>/gi, '<blockquote$1><p>');
+			// Removes spaces and switches the p tag and blockquote.
 			text = text.replace( /<\/blockquote>\s*<\/p>/gi, '</p></blockquote>');
+			// Removes spaces at the start and p tags from block level elements.
 			text = text.replace( new RegExp( '<p>\\s*(</?(?:' + blocklist + ')(?: [^>]*)?>)', 'gi' ), '$1' );
+			// Removes spaces at the end and p tags from block level elements.
 			text = text.replace( new RegExp( '(</?(?:' + blocklist + ')(?: [^>]*)?>)\\s*</p>', 'gi' ), '$1' );
-
-			// Remove redundant spaces and line breaks after existing <br /> tags
+			// Removes spaces and newlines after a br tag.
 			text = text.replace( /(<br[^>]*>)\s*\n/gi, '$1' );
 
-			// Create <br /> from the remaining line breaks
+			// Replaces spaces followed by a newline with a br tag followed by a new line.
 			text = text.replace( /\s*\n/g, '<br />\n');
 
+			// Removes whitespaces and a single br tag after a block level element.
 			text = text.replace( new RegExp( '(</?(?:' + blocklist + ')[^>]*>)\\s*<br />', 'gi' ), '$1' );
+			/*
+			 * Removes a br tag preceding white spaces followed by
+			 * p li div dl dd dt th pre td ul ol element..
```

Comment:
replace element.. with followed by one of these elements:

Also I don't know if it is necessary to name every element.